MapboxDirections/Match/MBMatchOptions.swift

```diff
@@ -0,0 +1,113 @@
+     
+     - parameter location: An array of `CLLocation` objects representing locations the route should try to match the road network against. The array should contain at least two locations (the source and destination) and at most 25 waypoints. (Some profiles, such as `MBDirectionsProfileIdentifierAutomobileAvoidingTraffic`, [may have lower limits](https://www.mapbox.com/api-documentation/#directions).)
+     - parameter profileIdentifier: A string specifying the primary mode of transportation for the routes. This parameter, if set, should be set to `MBDirectionsProfileIdentifierAutomobile`, `MBDirectionsProfileIdentifierAutomobileAvoidingTraffic`, `MBDirectionsProfileIdentifierCycling`, or `MBDirectionsProfileIdentifierWalking`. `MBDirectionsProfileIdentifierAutomobile` is used by default.
+     */
+    public init(locations: [CLLocation], profileIdentifier: MBDirectionsProfileIdentifier? = nil) {
+        let waypoints = locations.map {
+            Waypoint(location: $0)
+        }
+        self.timestamps = locations.map { $0.timestamp }
+        super.init(waypoints: waypoints, profileIdentifier: profileIdentifier)
+    }
+    
+    public init(coordinates: [CLLocationCoordinate2D], profileIdentifier: MBDirectionsProfileIdentifier? = nil) {
+        let waypoints = coordinates.map {
+            Waypoint(coordinate: $0)
+        }
+        super.init(waypoints: waypoints)
+    }
+    
+    
+    /**
+     If true, the input locations are re-sampled for improved map matching results. The default is  `false`.
+     */
+    @objc open var resampleTraces: Bool = false
+    
+    
+    /**
+     Timestamps corresponding to each coordinate provided in the request.
+     
+     There must be as many `timestamps` as there are coordinates in the request.
+     */
+    @objc private var timestamps: [Date]?
+    
+    
+    /**
+     An IndexSet of unique integers representing which coordinates should be treated as `Wapoints`.
+     */
+    @objc open var waypointIndices: IndexSet?
+    
+    @objc public required convenience init?(coder decoder: NSCoder) {
+        self.init(coder: decoder)
+        resampleTraces = decoder.decodeBool(forKey: "resampleTraces")
+        timestamps = decoder.decodeObject(of: [NSArray.self, NSDate.self], forKey: "timestamps") as? [Date]
+        waypointIndices = decoder.decodeObject(of: NSIndexSet.self, forKey: "waypointIndices") as IndexSet?
+    }
+    
+    @objc public override func encode(with coder: NSCoder) {
+        coder.encode(resampleTraces, forKey: "resampleTraces")
+        coder.encode(timestamps, forKey: "timestamps")
+        coder.encode(waypointIndices, forKey: "waypointIndices")
+    }
+    
+    @objc private var paramsThatOverlap: [String] = ["geometries", "radiuses", "steps", "overview", "timestamps", "annotations"]
+    
+    override internal var params: [URLQueryItem] {
+        var params = super.params.filter {
+            paramsThatOverlap.contains($0.name)
```

Comment:
Slightly annoying but if we do not remove these query params from `super.params`, the server errors.

Reply:
Noting, the removed params are still exposed on MatchOptions. If a developer sets one of the unused options, it will just a no-op.

Reply:
This is another good case for having RouteOptions and MatchOptions be siblings. Conceptually, it only makes sense for MatchOptions to inherit from RouteOptions if RouteOptions is a strict subset of MatchOptions.

Making RouteOptions and MatchOptions into siblings isn’t as scary as it sounds:

1. Rename RouteOptions to DirectionsOptions.
1. Make MatchOptions a subclass of DirectionsOptions.
1. Create a new RouteOptions that looks a lot like MatchOptions, except with routing-specific options (essentially anything that doesn’t overlap with MatchOptions).
1. Consider turning DirectionsOptions into a protocol with a default implementation.